controllers/storageclient/badgerdb/simple.go

```diff
@@ -126,23 +133,199 @@ func (cl Client) SetMetric(applicationName string, version int, signature, metri
 	key := getUserKey(applicationName, version, signature, user)
 
+	fmt.Println(key)
+
 	return cl.db.Update(func(txn *badger.Txn) error {
 		e := badger.NewEntry([]byte(key), []byte("true")).WithTTL(cl.additionalOptions.TTL)
 		err := txn.SetEntry(e)
 		return err
 	})
 }
 
-// GetSummaryMetrics gets a summary of all the metrics from all versions of an application
-func (cl Client) GetSummaryMetrics(applicationName string) (*map[int]storageclient.VersionMetricSummary, error) {
-	return nil, nil
+// getUserCount gets the number of users
+func (cl Client) getUserCount(applicationName string, version int, signature string) (uint64, error) {
+	count := uint64(0)
+
+	err := cl.db.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchValues = false
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		prefix := []byte(getUserPrefix(applicationName, version, signature))
+		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return count, nil
+}
+
+// GetSummaryMetrics gets summarized metrics for a particular application, version, and signature
+func (cl Client) GetSummaryMetrics(applicationName string, version int, signature string) (*storageclient.VersionMetricSummary, error) {
+	count, err := cl.getUserCount(applicationName, version, signature)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get user count: %e", err)
+	}
+
+	vms := storageclient.VersionMetricSummary{
+		NumUsers:        count,
+		MetricSummaries: map[string]storageclient.MetricSummary{},
+	}
+
+	// get metric summaries
+	// iter8 over all metrics with the provided application name, version, and signature
+	err = cl.db.View(func(txn *badger.Txn) error {
+		// used to capture all the metric values for the current metric
+		var overUserData []float64        // cumulative metric value for a particular user
+		var overTransactionData []float64 // all metric values
+
+		// used to determine what the previous user and metric are in the previous iteration
+		var previousUser string
+		var previousMetric string
+
+		it := txn.NewIterator(badger.DefaultIteratorOptions)
+		defer it.Close()
+
+		prefix := []byte(getMetricPrefix(applicationName, version, signature))
+		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+			item := it.Item()
+
+			// get key
+			key := string(item.Key())
+			err := item.Value(func(v []byte) error {
+				// get value
+				val := string(v)
+				fval, err := strconv.ParseFloat(val, 64)
+				if err != nil {
+					return fmt.Errorf("cannot convert metric value \"%s\" into float64", val)
+				}
+
+				// break down key into tokens
+				tokens := strings.Split(key[len(prefix):], "::")
+				if len(tokens) != 3 {
+					return fmt.Errorf("incorrect number of tokens in metric key \"%s\"; expected 7 tokens", key)
+				}
+
+				currentMetric := tokens[0]
+				currentUser := tokens[1]
+
+				// initial case
+				if previousMetric == "" {
```

Comment:
Is there a guarantee that users are going to be grouped together in this for loop?

What if you have:

```
user1
user2
user2
user1
user3
user2
user1
```

Reply:
According to the docs, yes. It says that keys are lexicographically sorted as you iterate through them.